showcase3D.py

```diff
@@ -0,0 +1,183 @@
+moves = {
+            # (axis, slice, dir)
+            "L": (0, 0, 1), "R": (0, 2, 1), "D": (1, 0, 1),
+            "U": (1, 2, 1), "B": (2, 0, 1), "F": (2, 2, 1),
+            "L'": (0, 0, -1), "R'": (0, 2, -1), "D'": (1, 0, -1),
+            "U'": (1, 2, -1), "B'": (2, 0, -1), "F'": (2, 2, -1),
+        }
+
+class Cubie3D():
+    def __init__(self, id, N, scale):
+        self.N = N
+        self.scale = scale
+        self.init_i = [*id]
+        self.current_i = [*id]
+        self.rot = [[1 if i==j else 0 for i in range(3)] for j in range(3)] # identity matrix
+
+    def reset(self):
+        self.current_i = [*self.init_i]
+        self.rot = [[1 if i==j else 0 for i in range(3)] for j in range(3)]
+
+    def isAffected(self, axis, slice, dir):
+        return self.current_i[axis] == slice
+
+    def update(self, axis, slice, dir):
+
+        if not self.isAffected(axis, slice, dir):
+            return
+
+        i, j = (axis+1) % 3, (axis+2) % 3
+        for k in range(3):
+            self.rot[k][i], self.rot[k][j] = -self.rot[k][j]*dir, self.rot[k][i]*dir
+
+        self.current_i[i], self.current_i[j] = (
+            self.current_i[j] if dir < 0 else self.N - 1 - self.current_i[j],
+            self.current_i[i] if dir > 0 else self.N - 1 - self.current_i[i] )
+
+    def transformMat(self):
+        scaleA = [[s*self.scale for s in a] for a in self.rot]  
+        scaleT = [(p-(self.N-1)/2)*2.1*self.scale for p in self.current_i] 
+        return [*scaleA[0], 0, *scaleA[1], 0, *scaleA[2], 0, *scaleT, 1]
+
+    def draw(self, animate, angle, axis, slice, dir):
+
+        glPushMatrix()
+
+        if animate and self.isAffected(axis, slice, dir):
+            glRotatef( angle*dir, *[1 if i==axis else 0 for i in range(3)] )
+        glMultMatrixf( self.transformMat() )
+
+        glBegin(GL_QUADS)
+        for i in range(len(centers)):
+            glColor3fv(colors[i])
+            for j in centers[i]:
+                glVertex3fv(vertices[j])
+        glEnd()
+
+        glPopMatrix()
+
+class Rubik3D():
+    def __init__(self, N, scale):
+        self.N = N
+        nbr = range(self.N)
+        self.cubes = [Cubie3D((x, y, z), self.N, scale) for x in nbr for y in nbr for z in nbr] # create the 27 cubies for a 3x3x3 Rubik
+
+    def mainloop(self, mix, solution):
+
+        rot_cube_map  = { K_UP: (-1, 0), K_DOWN: (1, 0), K_LEFT: (0, -1), K_RIGHT: (0, 1)} # rotate all rubik to change perspective view
+
+        ang_x, ang_y, rot_cube = 0, 0, (0, 0)
+        animate, animate_ang, animate_speed = False, 0, 5
+        action = (0, 0, 0)
+        current_move_index = 0
+
+        showing_solution = False
+        showing_mix = False
+
+        while True:
+
+            # KEY EVENTS
+
+            for event in pygame.event.get():
+                if event.type == pygame.QUIT:
+                    pygame.quit()
+                    quit()
+                if event.type == KEYDOWN:
```

Comment:
I think that the standard in GUIs is more to trigger the action at when the user releases the key but that's not very important point

Reply:
interesting! but for the arrows (changing cube orientation) it might be better to be able to press theoritically indefinitely the arrow to change orientation as we wish right? anyway that's something we can easily change before pushing the project

Reply:
okay, got you! since we are not sure to use this modelization for eval version let's let it this way and discuss all this on evaluation day